src/widget/wspinnybase.cpp

```diff
@@ -0,0 +1,658 @@
+#include "waveform/vsyncthread.h"
+#include "widget/moc_wspinny.cpp"
+#include "widget/wspinny.h"
+#include "wimagestore.h"
+
+// The SampleBuffers format enables antialiasing.
+WSpinnyBase::WSpinnyBase(
+        QWidget* parent,
+        const QString& group,
+        UserSettingsPointer pConfig,
+        VinylControlManager* pVCMan,
+        BaseTrackPlayer* pPlayer)
+        : WGLWidget(parent),
+          WBaseWidget(this),
+          m_group(group),
+          m_pConfig(pConfig),
+          m_pPlayPos(nullptr),
+          m_pVisualPlayPos(nullptr),
+          m_pTrackSamples(nullptr),
+          m_pTrackSampleRate(nullptr),
+          m_pScratchToggle(nullptr),
+          m_pScratchPos(nullptr),
+          m_pVinylControlSpeedType(nullptr),
+          m_pVinylControlEnabled(nullptr),
+          m_pSignalEnabled(nullptr),
+          m_pSlipEnabled(nullptr),
+          m_bShowCover(true),
+          m_dInitialPos(0.),
+          m_iVinylInput(-1),
+          m_bVinylActive(false),
+          m_bSignalActive(true),
+          m_iVinylScopeSize(0),
+          m_fAngle(0.0f),
+          m_dAngleCurrentPlaypos(-1),
+          m_dAngleLastPlaypos(-1),
+          m_fGhostAngle(0.0f),
+          m_dGhostAngleCurrentPlaypos(-1),
+          m_dGhostAngleLastPlaypos(-1),
+          m_iStartMouseX(-1),
+          m_iStartMouseY(-1),
+          m_iFullRotations(0),
+          m_dPrevTheta(0.),
+          m_dRotationsPerSecond(MIXXX_VINYL_SPEED_33_NUM / 60),
+          m_bClampFailedWarning(false),
+          m_bGhostPlayback(false),
+          m_pPlayer(pPlayer),
+          m_pCoverMenu(new WCoverArtMenu(this)),
+          m_pDlgCoverArt(new DlgCoverArtFullSize(this, pPlayer, m_pCoverMenu)) {
+#ifdef __VINYLCONTROL__
+    m_pVCManager = pVCMan;
+#else
+    Q_UNUSED(pVCMan);
+#endif // __VINYLCONTROL__
+    // Drag and drop
+    setAcceptDrops(true);
+    qDebug() << "WSpinnyBase(): Created WGLWidget, Context"
+             << "Valid:" << isContextValid()
+             << "Sharing:" << isContextSharing();
+    makeCurrentIfNeeded();
+
+    CoverArtCache* pCache = CoverArtCache::instance();
+    if (pCache) {
+        connect(pCache,
+                &CoverArtCache::coverFound,
+                this,
+                &WSpinnyBase::slotCoverFound);
+    }
+
+    if (m_pPlayer != nullptr) {
+        connect(m_pPlayer, &BaseTrackPlayer::newTrackLoaded, this, &WSpinnyBase::slotLoadTrack);
+        connect(m_pPlayer, &BaseTrackPlayer::loadingTrack, this, &WSpinnyBase::slotLoadingTrack);
+        // just in case a track is already loaded
+        slotLoadTrack(m_pPlayer->getLoadedTrack());
+    }
+
+    connect(m_pCoverMenu,
+            &WCoverArtMenu::coverInfoSelected,
+            this,
+            &WSpinnyBase::slotCoverInfoSelected);
+    connect(m_pCoverMenu, &WCoverArtMenu::reloadCoverArt, this, &WSpinnyBase::slotReloadCoverArt);
+}
+
+WSpinnyBase::~WSpinnyBase() {
+#ifdef __VINYLCONTROL__
+    m_pVCManager->removeSignalQualityListener(this);
+#endif
+}
```

Comment:
Can you convert the owned pointer members of the class to smart pointers? Otherwise they'll get leaked.

Reply:
note that this is not my code, but the original wspinny code, so if you think this leaks, it should be addressed in 2.3 / 2.4 / main as well.

Reply:
Yes, I'm aware its not your code. I just double-checked it and it doesn't leak because its using the QObject-based hierarchical memory management. I'll make that explicit in a future cleanup PR by making the members `parented_ptr`s. Let's postpone that until after merge this to avoid merge conflicts.

Reply:
Sorry for the false alarm.

Reply:
No problem at all! The mixture of QObject memory management and smart pointers is always tricky.